opm/simulators/gatherDeferredLogger.cpp

```diff
@@ -0,0 +1,173 @@
+            int offset = displ[process];
+            // unpack number of messages
+            unsigned int messagesize;
+            MPI_Unpack(data, recv_buffer.size(), &offset, &messagesize, 1, MPI_UNSIGNED, MPI_COMM_WORLD);
+            for (unsigned int i=0; i<messagesize; i++) {
+                messages.push_back(unpackSingleMessage(recv_buffer, offset));
+            }
+            assert(offset == displ[process + 1]);
+        }
+        return messages;
+    }
+
+} // anonymous namespace
+
+
+namespace Opm
+{
+
+    /// combine (per-process) messages
+    Opm::DeferredLogger gatherDeferredLogger(const Opm::DeferredLogger& local_deferredlogger)
+    {
+
+        int num_messages = local_deferredlogger.messages_.size();
+
+        int int64_mpi_pack_size;
+        MPI_Pack_size(1, MPI_INT64_T, MPI_COMM_WORLD, &int64_mpi_pack_size);
+        int unsigned_int_mpi_pack_size;
+        MPI_Pack_size(1, MPI_UNSIGNED, MPI_COMM_WORLD, &unsigned_int_mpi_pack_size);
+
+        // store number of messages;
+        int message_size = unsigned_int_mpi_pack_size;
+        // store 1 unsigned int for number of messages
+        message_size += num_messages*int64_mpi_pack_size;
+        // store 2 unsigned ints per message for length of tag and length of text
+        message_size += num_messages*2*unsigned_int_mpi_pack_size;
+
+        int string_mpi_pack_size;
+        for (const auto lm : local_deferredlogger.messages_) {
+            MPI_Pack_size(lm.tag.size(), MPI_CHAR, MPI_COMM_WORLD, &string_mpi_pack_size);
+            message_size += string_mpi_pack_size;
+            MPI_Pack_size(lm.text.size(), MPI_CHAR, MPI_COMM_WORLD, &string_mpi_pack_size);
+            message_size += string_mpi_pack_size;
+        }
+
+        // Pack local messages.
+        std::vector<char> buffer(message_size);
+
+        int offset = 0;
+        packMessages(local_deferredlogger.messages_, buffer, offset);
+        assert(offset == message_size);
+
+        // Get message sizes and create offset/displacement array for gathering.
+        int num_processes = -1;
+        MPI_Comm_size(MPI_COMM_WORLD, &num_processes);
+        std::vector<int> message_sizes(num_processes);
+        MPI_Allgather(&message_size, 1, MPI_INT, message_sizes.data(), 1, MPI_INT, MPI_COMM_WORLD);
+        std::vector<int> displ(num_processes + 1, 0);
+        std::partial_sum(message_sizes.begin(), message_sizes.end(), displ.begin() + 1);
+
+        // Gather.
+        std::vector<char> recv_buffer(displ.back());
+        MPI_Allgatherv(buffer.data(), buffer.size(), MPI_PACKED,
+                       const_cast<char*>(recv_buffer.data()), message_sizes.data(),
+                       displ.data(), MPI_PACKED,
+                       MPI_COMM_WORLD);
+
+        // Unpack.
+        std::vector<Opm::DeferredLogger::Message> m = unpackMessages(recv_buffer, displ);
+        Opm::DeferredLogger global_deferredlogger;
+        global_deferredlogger.messages_ = m;
```

Comment:
Instead of copying, either do std::move(), or (simpler, better) directly
```
global_deferredlogger.messages_ = unpackMessages(recv_buffer, displ);
```